### PR TITLE
[Feature] Add `optimize_anchors` script

### PR DESCRIPTION
**Pull request by imAzhou**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tools/analysis_tools/optimize_anchors.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+            prog_bar.update()
+        print('\n')
+        anchors = cluster_centers[np.argsort(
+            cluster_centers.prod(1))]  # sort small to large
+        self.logger.info(f'Anchor cluster finish. fitness = {f:.4f}')
+
+        return anchors
+
+    def metric(self, k, wh):  # compute metric
+        # wh（gt）的 shape 是 [N, 2] 表示 N 个 gt box 和宽高这 2 个维度
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:324]: No Chinese annotations

File: tools/analysis_tools/optimize_anchors.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+
+    def generate_gtwh_kwh(self):
+        r'''generate gt_bbox with width and height,
+           also return init cluster center anchor k
+           Return:
+               wh: Tensor(gt_nums, 2) the gt width and height
+                   after augment scale and small box filter
+               k: np.array(num_anchors, 2) kmeans evolved anchors
+        '''
+        # 根据均匀分布，随机生成0.9~1.1之间的缩放尺度
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:264]: No Chinese annotations

File: tools/analysis_tools/optimize_anchors.py [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+        '--input-shape',
+        type=int,
+        nargs='+',
+        default=[608, 608],
+        help='input image size')
+    parser.add_argument(
+        '--algorithm',
+        default='differential_evolution',
+        help='Algorithm used for anchor optimizing.'
+        'Support k-means and differential_evolution for YOLO.')
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:57]: Now there is one more option, and the corresponding description will be updated

File: tools/analysis_tools/optimize_anchors.py [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+    parser.add_argument(
+        '--iters',
+        default=1000,
+        type=int,
+        help='Maximum iterations for optimizer.')
+    parser.add_argument(
+        '--prior_match_thr',
+        default=4.0,
+        type=float,
+        help='anchor-label gt_filter_sizes ratio threshold hyperparameter.'
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:67]: Explain that this parameter is unique to `YOLOV5KMeansAnchorOptimizerter`

File: tools/analysis_tools/optimize_anchors.py [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+
+        thr = 1 / prior_match_thr
+
+        # step1: filter small bbox
+        box_size = self._filter_box(box_size)
+        assert num <= len(box_size)
+
+        # step2: init anchors
+        try:
+            from scipy.cluster.vq import kmeans
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:301]: 要提示用户： 由于 scipy 没有安装，目前采用随机初始化方式，如果你想得到更好的初始化，请安装 scipy，命令为： xxx。

还有这个 try 太大了，一旦出错用户无法判断是 scipy 没有安装，还是内部计算出错的。所以这个地方要优化。scipy try 放到最外面 import 时候，这里的 try 只是代码正确性的 try

File: tools/analysis_tools/optimize_anchors.py [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+        self.logger.info('init done, beginning evolve anchors...')
+        # sort small to large
+        anchors = torch.tensor(anchors[np.argsort(anchors.prod(1))]).to(
+            box_size.device, dtype=torch.float32)
+
+        # step3: evolve anchors
+        prog_bar = ProgressBar(iters)
+        fitness = self._anchor_fitness(box_size, anchors, thr)
+        cluster_shape = anchors.shape
+        mutation_prob, sigma = 0.9, 0.1
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:324]: 作为初始化参数，有默认值，但是可供用户修改

File: tools/analysis_tools/optimize_anchors.py [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+
+        for _ in range(iters):
+            mutate_result = np.ones(cluster_shape)
+            # mutate until a change occurs (prevent duplicates)
+            while (mutate_result == 1).all():
+                mutate_result = (
+                    (np.random.random(cluster_shape) < mutation_prob) *
+                    random.random() * np.random.randn(*cluster_shape) * sigma +
+                    1).clip(0.3, 3.0)
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:333]: 0.3 3 的含义是啥？

File: tools/analysis_tools/optimize_anchors.py [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+                is (anchor_num, 2), the number 2 means width and height.
+            threshold (float): the compare threshold of ratio
+
+        Returns:
+            Tuple: a tuple of metric result, best_ratio_mean and mean_matched
+        """
+        # step1: augment scale
+        # According to the uniform distribution,
+        # the scaling scale between 0.9 and 1.1 is randomly generated
+        scale = np.random.uniform(0.9, 1.1, size=(box_size.shape[0], 1))
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:368]: 0.9 1.1 也要作为初始化参数，提供morz

File: tools/analysis_tools/optimize_anchors.py [hunk 9]
```diff
@@ -0,0 +1,633 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+"""Optimize anchor settings on a specific dataset.
+
+This script provides two method to optimize YOLO anchors including k-means
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:4]:
```suggestion
This script provides three method to optimize YOLO anchors including k-means
```

File: tools/analysis_tools/optimize_anchors.py [hunk 10]
```diff
@@ -0,0 +1,633 @@
...
+anchor cluster and differential evolution. You can use ``--algorithm k-means``
+and ``--algorithm differential_evolution`` to switch two method.
+
+Example:
+    Use k-means anchor cluster::
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:9]: 三个例子

File: tools/analysis_tools/optimize_anchors.py [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+
+        python tools/analysis_tools/optimize_anchors.py ${CONFIG} \
+        --algorithm k-means --input-shape ${INPUT_SHAPE [WIDTH HEIGHT]} \
+        --output-dir ${OUTPUT_DIR}
+    Use differential evolution to optimize anchors::
```
Comment (hhaAndroid) [inline tools/analysis_tools/optimize_anchors.py:14]: 可以在 docs/zh_cn/user_guides/useful_tools.md 加一下用法吗？ 中英文都要。可以以气球数据集为例子给个具体例子，并贴一下具体结果，三个可选算法都要给例子。

**Comment by hhaAndroid** [comment]
请同步代码，并修复冲突和 ci

File: docs/zh_cn/user_guides/useful_tools.md [hunk 12]
```diff
@@ -231,6 +231,44 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
 
 转换好的 `mmyolox.pt` 即可以在 MMYOLO 中使用。
 
+## 优化锚框尺寸
+
+脚本 `tools/analysis_tools/optimize_anchors.py` 支持yolo系列中三种锚框生成方式，分别是 `k-means`、`differential_evolution`、`v5-k-means`.
```
Comment (hhaAndroid) [inline docs/zh_cn/user_guides/useful_tools.md:236]: 持yolo系列
```suggestion
脚本 `tools/analysis_tools/optimize_anchors.py` 支持 yolo 系列中三种锚框生成方式，分别是 `k-means`、`differential_evolution`、`v5-k-means`.
```

File: docs/zh_cn/user_guides/useful_tools.md [hunk 13]
```diff
@@ -231,6 +231,44 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
...
+
+### k-means
+
+在 k-means 方法中，使用的是基于IoU表示距离的聚类方法，具体使用命令如下:
```
Comment (hhaAndroid) [inline docs/zh_cn/user_guides/useful_tools.md:240]:
```suggestion
在 k-means 方法中，使用的是基于 IoU 表示距离的聚类方法，具体使用命令如下:
```

File: docs/zh_cn/user_guides/useful_tools.md [hunk 14]
```diff
@@ -231,6 +231,44 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
...
+```shell
+python tools/analysis_tools/optimize_anchors.py ${CONFIG} \
+    --algorithm differential_evolution \
+    --input-shape ${INPUT_SHAPE [WIDTH HEIGHT]} \
+    --output-dir ${OUTPUT_DIR}
+```
+
+### v5-k-means
+
+在 v5-k-means 方法中，使用的是yolov5中基于shape-match的聚类方式，具体使用命令如下:
```
Comment (hhaAndroid) [inline docs/zh_cn/user_guides/useful_tools.md:262]:
```suggestion
在 v5-k-means 方法中，使用的是 yolov5 中基于 shape-match 的聚类方式，具体使用命令如下:
```

**Comment by OpenMMLab-Miao** [comment]
Dear imAzhou,
First of all, we want to express our gratitude for your significant PR in the MMYOLO project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.
If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. We look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG
Thank you again for your contribution❤
Best regards！ @imAzhou

**Pull request merged**